src/__tests__/applicants.int.test.ts

```diff
@@ -66,6 +66,33 @@ describe('/applicants', () => {
         );
     });
 
+    it('returns the expected applicant by externalId', async () => {
+      await db.query(`
+        INSERT INTO applicants (
+          external_id,
+          opted_in,
+          created_at
+        )
+        VALUES
+          ( '54280485', 'false', '2023-04-25 17:39:00+0000' ),
+          ( '167336', 'true', '2023-04-05 17:40:00+0000' );
+      `);
+      await agent
+        .get('/applicants?externalId=167336')
+        .set(authHeader)
+        .expect(
+          200,
+          [
+            {
+              createdAt: '2023-04-05T17:40:00.000Z',
+              externalId: '167336',
+              id: 2,
+              optedIn: true,
+            },
+          ],
+        );
```

Comment:
Here is the reasoning:
1. We want a consistent API.
2. We consistently use the `GET /[entity]/{id}` path parameter as PDC-assigned/internal/DB-generated ID. This is guaranteed to return one result in all entities.
3. External ID on applicants is not the same as a PDC-assigned/internal/DB-generated ID. Therefore it is a query parameter.
4. We (will) use query parameters for filtering/searching across all entities.
5. When filtering/searching entities by a query parameter, in general, the result will be many. This particular result happens to be one instead of many.
6. For consistency, we want the response shape to be the same for `GET /applicants` whether we filter or do not filter. We don't want the shape of the response to change with query parameters, just the amount of data or in some narrow cases an optional field present or not present.
7. When a filter/search comes up lacking, one typically would get an empty list instead.

So I agree, but I also already considered that in light of the above points.

Reply:
I agree on 1-4! However, I don't think 5 or 7 applies, and disagree on 6. Returning a single object or `404`ing IS being consistent - just with `/applicants/:id` rather than with `/applicants?_search={term}`. This, I suppose, is the benefit of separating out `/entities` from `/entity/{id}`, because under that naming convention, this API would be `/applicant?externalId={id}`, making it clear that one and only one result is possible.

Query parameters can be a part of stable URLs, and I don't think we should treat all query parameters the same. A zero- or single-element list is a more cumbersome API to consume than one that gives an object or a 404. Let's do the thing that makes our API easier to consume!

Reply:
For what it's worth I also would probably interpret attribute-level query params as a filter / search and would expect the response to be the same as other filters in the event of zero matches.  Which is to say, I would be surprised to get a 404 because we provided a search facet that resulted in zero results, and I would expect an empty array.

If we changed to `/applicant?externalId=...` would signal a specific resource attempt, but I also think that breaks my expectations around REST resource names (I expect the resource category to be the same `/applicants` for both single and collection access).  I'm happy to be told my expectations are out of date, but I think they're rooted in at least some historic form of best practice...

If we really want the 404 behavior, what about What about something like `/applicants/externalId/{foo}` which is described [here](https://stackoverflow.com/questions/56969324/rest-api-resource-naming-with-multiple-ids).

The downside of this approach is we now have two distinct URLs for the same resource, whereas with the "filter" approach the collection returned is not pointing to a resource but rather is returning a collection of resources and there is still only one resource URL.

Reply:
(oh, I do think this discussion signals that a test case would be reasonable though, to make it clear the intended implementation)

Reply:
In short, @jasonaowen, I would be surprised or given pause by an API that sometimes returns a single result for query parameters and other times returns a list for query parameters, so I do not think of such an API as easier to consume. I added the test suggested by @slifty.

Reply:
Okay, it seems the main difference is that you are both framing this as a search rather than a lookup. Under that framing, returning an array makes sense!

Reserving the right to later make querying only by `externalId` return more than one response is also a convincing argument, although it's not obvious to me what that'd mean for API versioning/semver - would that be a breaking change? Not something we have to figure out yet, either way.

Reply:
@jasonaowen Not to beat a dead horse but to acknowledge that this entity is different: in the original API design, I think this was an endpoint that used the natural key (`externalId`) as the URL parameter rather than the synthetic key, but that was an exception among other entities. So we made this one consistent with the synthetic key as a URL parameter. And then yes, I think both @slifty and I think we should consistently treat query parameters as search parameters.